File: app/modules/simulator.py
"""Tab 3 -- Mock-Draft Simulator. Auto-sim or go on-the-clock for one team."""
from __future__ import annotations

import io

import pandas as pd
from shiny import module, reactive, render, ui

from logic import dataio, github_storage
from logic.sim_engine import simulate

# First-round mock: only the top 50 consensus players are in the draftable pool.
_BOARD = (dataio.consensus().sort_values("consensus_rank").head(50).to_dict("records")
          if len(dataio.consensus()) else [])
_ORDER = dataio.draft_order()
_TEND = dataio.team_tendencies()
_TEAMS = [s.get("team", "") for s in _ORDER]
_BY_ID = {p["player_id"]: p for p in _BOARD}


def _choice_label(p: dict) -> str:
    return f"#{int(p['consensus_rank'])}  {p['player']} ({p.get('position','')}, {p.get('school','')})"


@module.ui
def simulator_ui():
    if not _BOARD or not _ORDER:
        return ui.p("Simulator needs both a consensus board and a 2026 draft order -- "
                    "run the scraper first.", class_="muted")
    return ui.layout_sidebar(
        ui.sidebar(
            ui.input_select("mode", "Engine", {
                "realistic": "Realistic (team lean + noise)",
                "team_need": "Team need (no noise)",
                "bpa": "Best player available",
            }, selected="realistic"),
            ui.input_slider("rand", "Unpredictability", 0.0, 0.4, 0.15, step=0.05),
            ui.input_numeric("seed", "Random seed", 1, min=1, max=99999),
            ui.hr(),
            ui.input_select("team", "Go on the clock as", ["(spectate -- auto only)"] + _TEAMS),
            ui.input_action_button("reset", "Reset draft", class_="btn-sm"),
            ui.download_button("dl", "Download CSV", class_="btn-sm"),
            ui.hr(),
            ui.input_action_button("save_btn", "Save Draft", class_="btn-sm"),
            width=320,
        ),
        ui.output_ui("clock"),
        ui.output_ui("pick_panel"),
        ui.output_ui("results"),
    )


@module.server
def simulator_server(input, output, session):
    committed = reactive.value({})  # pick_no -> player_id (user + finalized CPU)
    save_status = reactive.value("")

    @reactive.effect
    @reactive.event(input.reset, input.mode, input.rand, input.seed)
    def _reset():
        committed.set({})

    @reactive.calc
    def projection() -> list[dict]:
        """Full projected draft given current committed picks."""
        if not _BOARD or not _ORDER:
            return []
        return simulate(_BOARD, _ORDER, _TEND, mode=input.mode(),
                        randomness=input.rand(), seed=int(input.seed() or 1),
                        locked=committed())

    def _on_clock_idx() -> int | None:
        c = committed()
        for i, slot in enumerate(_ORDER):
            if slot.get("pick") not in c:
                return i
        return None

    @render.ui
    def clock():
        idx = _on_clock_idx()
        if idx is None:
            return ui.div(ui.h3("Draft complete"), class_="clock-done")
        slot = _ORDER[idx]
        my_team = input.team()
        on_user = my_team in _TEAMS and slot.get("team") == my_team

        head = ui.h3(f"Pick {slot.get('pick')} -- {slot.get('team')} "
                     f"{'(you)' if on_user else 'on the clock'}")
        if on_user:
            return ui.div(
                head,
                ui.input_text("pick_search", None,
                              placeholder="Search players by name, school, or position…",
                              width="100%"),
                ui.output_text_verbatim("pick_selection", placeholder=True),
                ui.input_action_button("draft", "Draft selected", class_="btn-primary"),
                class_="clock-box",
            )
        return ui.div(
            head,
            ui.input_action_button("advance", "Auto-pick this one", class_="btn-sm"),
            ui.input_action_button("tonext", "Sim to my next pick", class_="btn-sm"),
            ui.input_action_button("auto", "Auto-complete draft", class_="btn-primary"),
            class_="clock-box",
        )

    # -- Player search for on-clock picks -----------------------------------

    selected_pid: reactive.Value[str] = reactive.Value("")
    input_id = session.ns("pickfor")

    @reactive.calc
    def _avail() -> list[dict]:
        taken = set(committed().values())
        return [p for p in _BOARD if p["player_id"] not in taken]

    @render.ui
    def pick_panel():
        idx = _on_clock_idx()
        if idx is None:
            return ui.div()
        slot = _ORDER[idx]
        my_team = input.team()
        if not (my_team in _TEAMS and slot.get("team") == my_team):
            return ui.div()
        q = (input.pick_search() or "").strip().lower()
        avail = _avail()
        if q:
            avail = [p for p in avail
                     if q in p["player"].lower()
                     or q in p.get("school", "").lower()
                     or q in p.get("position", "").lower()][:20]
        else:
            avail = avail[:20]
        sel = selected_pid()
        rows = []
        for p in avail:
            pid = p["player_id"]
            is_sel = pid == sel
            safe_pid = pid.replace("'", "\\'")
            style = ("background:var(--accent-bg);border-color:var(--accent);"
                     if is_sel else "")
            rows.append(
                f"<div class='sim-pick-row' style='{style}' "
                f"onclick=\"Shiny.setInputValue('{input_id}', '{safe_pid}', {{priority:'event'}});\">"
                f"<span class='bb-rank'>#{int(p['consensus_rank'])}</span> "
                f"<strong>{p['player']}</strong> "
                f"<span class='bb-pos'>{p.get('position','')} · {p.get('school','')}</span>"
                f"</div>"
            )
        if not rows:
            return ui.div(ui.p("No players found.", class_="muted"),
                          style="margin-top:.5rem;")
        return ui.div(
            ui.HTML("<div class='sim-pick-list'>" + "".join(rows) + "</div>"),
            style="margin-top:.5rem;",
        )

    @reactive.effect
    @reactive.event(input.pickfor)
    def _on_pick_select():
        pid = input.pickfor()
        if pid:
            selected_pid.set(pid)

    @render.text
    def pick_selection():
        pid = selected_pid()
        if not pid:
            return "No player selected"
        p = next((x for x in _BOARD if x["player_id"] == pid), None)
        if not p:
            return "No player selected"
        return f"Selected: #{int(p['consensus_rank'])} {p['player']} ({p.get('position','')})"

    @reactive.effect
    @reactive.event(input.draft)
    def _draft():
        idx = _on_clock_idx()
        if idx is None:
            return
        pid = selected_pid()
        if not pid:
            return
        c = dict(committed())
        c[_ORDER[idx].get("pick")] = pid
        selected_pid.set("")
        committed.set(c)

    @reactive.effect
    @reactive.event(input.advance)
    def _advance():
        _commit_cpu_until(stop_at_user=False, one=True)

    @reactive.effect
    @reactive.event(input.tonext)
    def _tonext():
        _commit_cpu_until(stop_at_user=True, one=False)

    @reactive.effect
    @reactive.event(input.auto)
    def _auto():
        proj = {r["pick"]: r["player_id"] for r in projection()}
        committed.set(proj)

    def _commit_cpu_until(*, stop_at_user: bool, one: bool):
        proj = {r["pick"]: r["player_id"] for r in projection()}
        c = dict(committed())
        my_team = input.team()
        for slot in _ORDER:
            pk = slot.get("pick")
            if pk in c:
                continue
            if slot.get("team") == my_team and my_team in _TEAMS:
                break  # reached the user's pick
            c[pk] = proj.get(pk)
            if one:
                break
            if not stop_at_user:
                break
        committed.set(c)

    @render.ui
    def results():
        proj = projection()
        if not proj:
            return ui.div()
        my_team = input.team()
        rows = []
        cset = committed()
        for r in proj:
            done = r["pick"] in cset
            tag = ("you" if r["team"] == my_team and my_team in _TEAMS else "")
            cls = " sim-you" if tag else ""
            cls += " sim-locked" if done else " sim-proj"
            val = r["value"]
            # val = consensus_rank - pick; negative = value (player fell), positive = reach
            vtxt = (f'<span class="val-neg">+{val}</span>' if isinstance(val, (int, float)) and val > 0
                    else (f'<span class="val-pos">{val}</span>' if isinstance(val, (int, float)) and val < 0 else ""))
            rows.append(
                f'<tr class="{cls}"><td>{r["pick"]}</td><td>{r["team"]}</td>'
                f'<td>{r["player"]}</td><td>{r["position"]}</td>'
                f'<td>{r["school"]}</td><td>{r["consensus_rank"]}</td><td>{vtxt}</td></tr>'
            )
        return ui.div(
            ui.help_text("Solid rows are finalized; faded rows are projected. "
                         "Green = value (player fell); Red = reach (taken early)."),
            ui.HTML('<div class="sim-wrap"><table class="sim-table"><thead><tr>'
                    '<th>Pick</th><th>Team</th><th>Player</th><th>Pos</th><th>School</th>'
                    '<th>Cons.</th><th>Value</th></tr></thead><tbody>'
                    + "".join(rows) + "</tbody></table></div>"),
        )

    # -- Save Draft ----------------------------------------------------------

    @reactive.effect
    @reactive.event(input.save_btn)
    def _show_save_modal():
        save_status.set("")
        m = ui.modal(
            ui.input_text("draft_name", "Draft name", placeholder="My Mock Draft"),
            ui.input_text("draft_author", "Your name", placeholder="Your name"),
            ui.output_ui("save_msg"),
            title="Save Mock Draft",
            easy_close=True,
            footer=ui.div(
                ui.input_action_button("save_confirm", "Save", class_="btn-primary"),
                ui.modal_button("Cancel"),
            ),
        )
        ui.modal_show(m)

    @render.ui
    def save_msg():
        msg = save_status()
        if not msg:
            return ui.div()
        color = "var(--good)" if "saved" in msg.lower() else "var(--bad)"
        return ui.p(msg, style=f"color:{color};margin-top:.5rem;")

    @reactive.effect
    @reactive.event(input.save_confirm)
    async def _do_save():
        picks = projection()
        if not picks:
            save_status.set("No picks to save -- run the draft first.")
            return
        save_status.set("Saving...")
        safe_keys = ("pick", "team", "player", "position", "school", "consensus_rank")
        clean_picks = [{k: r[k] for k in safe_keys if k in r} for r in picks]
        ok, msg = await github_storage.save_draft(
            draft_name=input.draft_name(),
            author=input.draft_author(),
            mode=input.mode(),
            seed=int(input.seed() or 1),
            picks=clean_picks,
        )
        save_status.set(msg)
        if ok:
            ui.modal_remove()

    # -- Download ------------------------------------------------------------

    @render.download(filename="mock_draft_2026.csv")
    def dl():
        proj = projection()
        if proj:
            buf = io.StringIO()
            pd.DataFrame(proj)[["pick", "team", "player", "position", "school",
                                "consensus_rank", "value"]].to_csv(buf, index=False)
            yield buf.getvalue()

File: static/js/simulator.js
import { saveDraft } from './storage.js';

function esc(s) {
  return String(s??'').replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;').replace(/"/g,'&quot;');
}

// Mulberry32 seeded RNG (reproducible from seed)
function mulberry32(seed) {
  let s = seed >>> 0;
  return function() {
    s = (s + 0x6D2B79F5) >>> 0;
    let t = Math.imul(s ^ (s >>> 15), 1 | s);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function uniform(rng, lo, hi) { return lo + rng() * (hi - lo); }

function posGroup(pos) {
  pos = (pos || '').toUpperCase();
  if (['RHP','LHP','P'].includes(pos)) return 'P';
  if (pos === 'C') return 'C';
  if (pos === 'OF') return 'OF';
  return 'IF';
}

function needFactor(tend, player, takenGroups) {
  if (!tend) return 1.0;
  const grp = posGroup(player.position || '');
  const isPitcher = grp === 'P';
  const pctP = parseFloat(tend.pct_pitcher || 0.5) || 0.5;
  const lean = isPitcher ? pctP : (1.0 - pctP);
  let factor = 1.0 - 0.18 * (lean - 0.5);
  const pb = tend.position_breakdown || {};
  const total = Object.values(pb).reduce((a,b) => a+b, 0) || 1;
  const grpShare = (pb[grp] || 0) / total;
  factor *= 1.0 - 0.12 * (grpShare - 0.25);
  factor *= 1.0 + 0.06 * (takenGroups[grp] || 0);
  return factor;
}

function simulate(board, order, tendencies, { mode='realistic', randomness=0.15, seed=0, locked={} } = {}) {
  const rng = mulberry32(seed);
  const avail = new Map(board.map(p => [p.player_id, p]));
  const takenGroups = {};
  const results = [];

  for (const slot of order) {
    const pickNo = slot.pick, team = slot.team || '';
    const forcedId = locked[pickNo] || locked[String(pickNo)];
    let chosen = null, isManual = false;

    if (forcedId && avail.has(forcedId)) {
      chosen = avail.get(forcedId);
      isManual = true;
    } else if (avail.size > 0) {
      const tg = (takenGroups[team] = takenGroups[team] || {});
      const scored = [];
      for (const p of avail.values()) {
        const cr = p.consensus_rank || 9999;
        let score = cr;
        if (mode !== 'bpa') {
          score = cr * needFactor(tendencies[team], p, tg);
          if (mode === 'realistic' && randomness > 0) {
            score *= 1.0 + uniform(rng, -randomness, randomness);
          }
        }
        scored.push([score, p]);
      }
      scored.sort((a,b) => a[0]-b[0]);
      chosen = scored[0][1];
    }

    if (!chosen) continue;
    avail.delete(chosen.player_id);
    const grp = posGroup(chosen.position || '');
    const tg = (takenGroups[team] = takenGroups[team] || {});
    tg[grp] = (tg[grp] || 0) + 1;

    const cr = chosen.consensus_rank;
    results.push({
      pick: pickNo, team,
      player: chosen.player || '',
      position: chosen.position || '',
      school: chosen.school || '',
      class_level: chosen.class_level || '',
      consensus_rank: cr,
      value: (typeof cr === 'number' && pickNo) ? (cr - pickNo) : null,
      player_id: chosen.player_id,
      manual: isManual,
    });
  }
  return results;
}

export function initSimulator(G, showModal) {
  const el = document.getElementById('tab-simulator');
  const order = G.draft_order;
  const teams = [...new Set(order.map(s => s.team))].sort();
  const defaultSeed = Math.floor(Math.random() * 100000);

  el.innerHTML = `
    <div style="display:flex;gap:1.5rem;flex-wrap:wrap;align-items:flex-end;margin-bottom:1rem;">
      <div>
        <label class="control-label">Your team</label>
        <select id="sim-team" class="form-select" style="width:220px;">
          <option value="">(none — full sim)</option>
          ${teams.map(t => `<option>${esc(t)}</option>`).join('')}
        </select>
      </div>
      <div>
        <label class="control-label">Mode</label>
        <select id="sim-mode" class="form-select" style="width:160px;">
          <option value="realistic">Realistic</option>
          <option value="bpa">Best Player Available</option>
          <option value="team_need">Team Need</option>
        </select>
      </div>
      <div>
        <label class="control-label">Randomness</label>
        <input type="range" id="sim-rand" min="0" max="0.5" step="0.05" value="0.15" style="width:120px;vertical-align:middle;margin-left:.4rem;">
        <span id="sim-rand-val" style="font-family:'JetBrains Mono',monospace;margin-left:.3rem;">0.15</span>
      </div>
      <div>
        <label class="control-label">Seed</label>
        <input type="number" id="sim-seed" class="form-control" value="${defaultSeed}" style="width:110px;">
      </div>
      <button id="sim-start" class="btn btn-primary" style="align-self:flex-end;">Start Simulation</button>
    </div>
    <div id="sim-clock"></div>
    <div id="sim-results"></div>
    <div id="sim-save-area" style="display:none;margin-top:1rem;"></div>
  `;

  el.querySelector('#sim-rand').addEventListener('input', e => {
    el.querySelector('#sim-rand-val').textContent = e.target.value;
  });

  let state = null; // {order, committed, myTeam, mode, randomness, seed}

  function getOpts() {
    return {
      myTeam: el.querySelector('#sim-team').value,
      mode: el.querySelector('#sim-mode').value,
      randomness: parseFloat(el.querySelector('#sim-rand').value),
      seed: parseInt(el.querySelector('#sim-seed').value, 10) || 0,
    };
  }

  function nextMyPick(results, committed, myTeam) {
    if (!myTeam) return null;
    for (const slot of order) {
      if (slot.team !== myTeam) continue;
      if (committed[slot.pick] || committed[String(slot.pick)]) continue;
      // check if this slot has been reached in results
      const done = results.find(r => r.pick === slot.pick);
      if (!done) return slot.pick;
    }
    return null;
  }

  function renderClock(pickNo, results, committed) {
    const clockEl = document.getElementById('sim-clock');
    const avail = G.consensus.filter(p => !results.some(r => r.player_id === p.player_id));

    clockEl.innerHTML = `
      <div class="clock-box">
        <div style="font-size:.78rem;color:var(--muted);text-transform:uppercase;letter-spacing:.8px;margin-bottom:.3rem;">On the Clock — Pick #${pickNo}</div>
        <div style="margin-bottom:.6rem;">
          <input type="text" id="sim-search" class="form-control" placeholder="Search players…" autocomplete="off" style="max-width:340px;">
        </div>
        <div id="sim-pick-list" class="sim-pick-list"></div>
        <div style="margin-top:.7rem;display:flex;align-items:center;gap:.7rem;">
          <span id="sim-pick-name" style="font-size:.88rem;color:var(--accent);min-width:160px;"></span>
          <button id="sim-draft-btn" class="btn btn-primary btn-sm" disabled>Draft Player</button>
        </div>
      </div>`;

    let selectedPid = null;
    const listEl = document.getElementById('sim-pick-list');
    const searchEl = document.getElementById('sim-search');
    const nameEl = document.getElementById('sim-pick-name');
    const draftBtn = document.getElementById('sim-draft-btn');

    function renderList() {
      const q = searchEl.value.trim().toLowerCase();
      const top = avail.filter(p => !q || p.player.toLowerCase().includes(q) || (p.school||'').toLowerCase().includes(q)).slice(0,30);
      listEl.innerHTML = top.map(p => `
        <div class="sim-pick-row${p.player_id === selectedPid ? ' selected' : ''}" data-pid="${esc(p.player_id)}"
          style="${p.player_id === selectedPid ? 'background:var(--accent-bg);border-color:var(--accent);' : ''}">
          <span style="font-family:'JetBrains Mono',monospace;color:var(--accent);margin-right:.5rem;">#${p.consensus_rank}</span>
          <strong>${esc(p.player)}</strong>
          <span class="muted"> · ${esc(p.position)} · ${esc(p.school)}</span>
        </div>`).join('');
      listEl.querySelectorAll('[data-pid]').forEach(row => {
        row.addEventListener('click', () => {
          selectedPid = row.dataset.pid;
          const p = G.consensus.find(x => x.player_id === selectedPid);
          nameEl.textContent = p ? p.player : '';
          draftBtn.disabled = false;
          renderList();
        });
      });
    }
    searchEl.addEventListener('input', renderList);
    renderList();

    draftBtn.addEventListener('click', () => {
      if (!selectedPid) return;
      state.committed[pickNo] = selectedPid;
      runSim();
    });
  }

  function runSim() {
    const results = simulate(G.consensus, order, G.tendencies, {
      mode: state.mode,
      randomness: state.randomness,
      seed: state.seed,
      locked: state.committed,
    });

    const myPick = nextMyPick(results, state.committed, state.myTeam);

    if (myPick) {
      // Run sim up to (but not including) myPick
      const partial = results.filter(r => r.pick < myPick);
      renderResults(partial, true);
      renderClock(myPick, partial, state.committed);
    } else {
      renderResults(results, false);
      document.getElementById('sim-clock').innerHTML = results.length
        ? `<p class="clock-done" style="margin:.5rem 0;">Draft complete!</p>` : '';
      renderSaveArea(results);
    }
  }

  function renderResults(results, partial) {
    const resEl = document.getElementById('sim-results');
    if (!results.length) { resEl.innerHTML = ''; return; }
    const rows = results.map(r => {
      const isMe = r.team === state.myTeam;
      const valClass = r.value === null ? '' : r.value > 0 ? 'val-neg' : r.value < 0 ? 'val-pos' : '';
      const valStr = r.value === null ? '' : (r.value > 0 ? `+${r.value}` : String(r.value));
      return `<tr class="${isMe ? 'sim-you' : ''}${r.manual ? ' sim-locked' : ' sim-proj'}">
        <td style="font-family:'JetBrains Mono',monospace;color:var(--accent);">${r.pick}</td>
        <td class="muted" style="font-size:.8rem;text-align:left;">${esc(r.team)}</td>
        <td style="text-align:left;cursor:pointer;font-weight:600;" data-pid="${esc(r.player_id)}">${esc(r.player)}</td>
        <td class="bb-pos">${esc(r.position)}</td>
        <td class="muted" style="font-size:.8rem;text-align:left;">${esc(r.school)}</td>
        <td class="muted">${r.consensus_rank || ''}</td>
        <td class="${valClass}" style="font-family:'JetBrains Mono',monospace;">${valStr}</td>
      </tr>`;
    }).join('');
    resEl.innerHTML = `<div class="sim-wrap"><table class="sim-table">
      <thead><tr><th>Pick</th><th>Team</th><th>Player</th><th>Pos</th><th>School</th><th>Cons.</th><th>Value</th></tr></thead>
      <tbody>${rows}</tbody></table></div>`;
    resEl.querySelectorAll('[data-pid]').forEach(td => {
      td.addEventListener('click', () => showModal(td.dataset.pid));
    });
  }

  function renderSaveArea(results) {
    if (!results.length) return;
    const area = document.getElementById('sim-save-area');
    area.style.display = '';
    area.innerHTML = `
      <hr>
      <div style="display:flex;gap:.8rem;flex-wrap:wrap;align-items:flex-end;">
        <div>
          <label class="control-label">Draft name</label>
          <input type="text" id="sim-save-name" class="form-control" placeholder="My 2026 Mock Draft" style="width:240px;">
        </div>
        <div>
          <label class="control-label">Author</label>
          <input type="text" id="sim-save-author" class="form-control" placeholder="Your name" style="width:160px;">
        </div>
        <button id="sim-save-btn" class="btn btn-primary btn-sm">Save Draft</button>
        <span id="sim-save-status" style="font-size:.82rem;color:var(--muted);"></span>
      </div>`;

    document.getElementById('sim-save-btn').addEventListener('click', async () => {
      const name = document.getElementById('sim-save-name').value.trim();
      const author = document.getElementById('sim-save-author').value.trim();
      if (!name) { alert('Please enter a draft name.'); return; }
      const btn = document.getElementById('sim-save-btn');
      btn.disabled = true;
      btn.textContent = 'Saving…';
      try {
        await saveDraft({
          draftName: name,
          author: author || 'Anonymous',
          mode: state.mode,
          seed: state.seed,
          picks: results.map(r => ({ pick: r.pick, team: r.team, player_id: r.player_id, player: r.player, position: r.position })),
        });
        document.getElementById('sim-save-status').textContent = 'Draft saved!';
        btn.textContent = 'Saved';
      } catch(e) {
        document.getElementById('sim-save-status').textContent = 'Error: ' + e.message;
        btn.disabled = false;
        btn.textContent = 'Save Draft';
      }
    });
  }

  el.querySelector('#sim-start').addEventListener('click', () => {
    const opts = getOpts();
    state = { ...opts, committed: {} };
    document.getElementById('sim-save-area').style.display = 'none';
    runSim();
  });
}
